src/realm/db.cpp

```diff
@@ -759,7 +797,46 @@ class DB::FileVersionManager : public DB::VersionManager {
     }
 
     File& m_file;
-    File::Map<SharedInfo> m_reader_map;
+    File::Map<DB::SharedInfo> m_reader_map;
+
+    friend class DB::EncryptionMarkerObserver;
+};
+
+// adapter class for marking/observing encrypted writes
+class DB::EncryptionMarkerObserver : public util::WriteMarker, public util::WriteObserver {
+public:
+    EncryptionMarkerObserver(DB::VersionManager& vm)
+        : vm(vm)
+    {
+    }
+    bool no_concurrent_writer_seen() override
+    {
+        uint64_t tmp_write_count;
+        auto page_may_have_been_written = vm.observe_writer(nullptr, &tmp_write_count);
+        if (tmp_write_count != last_seen_count) {
+            page_may_have_been_written = true;
+            last_seen_count = tmp_write_count;
+        }
+        if (page_may_have_been_written) {
+            calls_since_last_writer_observed = 0;
+            return false;
+        }
+        ++calls_since_last_writer_observed;
+        return (calls_since_last_writer_observed >= 5);
```

Comment:
What's with the 5 here? Is it just an arbitrary maximum number of times to retry?